### DISCUSSION TITLE
Rabbitmq Operator deploys rabbitmq cluster as 2 clusters

### DISCUSSION THREAD

**Discussion by xixiangzouyibian** [rabbitmq/cluster-operator#1472] [Q&A] [answered]
I deployed 3-node rabbitmq cluster by rabbitmq operator, which specified replicas as 3. But it finally formed 2 rabbitmq clusters and they didn't recognize each other (empty list in network partitions). The detail of observations:
- These clusters have the same erlang cookie.
- Both 2 clusters could find all of peer nodes.
- The created quorum queues were randomly assinged to two clusters, but only one of clusters is working at the same time, so some of business are blocked due to the quorum queues ware not found and the messages can not be sent successfully. 

The split-brain behavior I understand is generate a cluster of 3 nodes at the beginning, but the cluster is split into 2 due to network partiion failure. However, in my case, the deployment from the beginning is two clusters. Only one of clusters provide the services, and the other one appears idle. 

How can I do for this case or how to mitigate this senario, and is it possible to restore the cluster as one ?

**Comment by xixiangzouyibian** [comment]
The phenomenon occurs frequently in our pipeline deployment, not sure if it is a rabbitmq or operator issue. 
Current rabbitmq version: 3.11.18
Erlang version: 25.3.2 [jit]
k8s: 
- Client Version: v1.26.4+rke2r1
- Kustomize Version: v4.5.7
- Server Version: v1.26.4+rke2r1

**Reply by mkuratczyk** [reply]
We hear about this problem occasionally, but are unable to reproduce / pin-point what triggers this. There are thousands of clusters managed by the Operator (and I personally deployed hundreds if not thousands for testing purposes) and we haven't seen this problem since the locks were introduced. The very low frequency of reports suggests it affects very few users.

This is not to say there is no problem, but we need help from someone who experiences the problem - at the very least, full logs from all nodes at debug level (and ideally on the latest 3.12 if you can try that).

see also:
https://github.com/rabbitmq/cluster-operator/issues/662#issuecomment-1112423621

**Reply by xixiangzouyibian** [reply]
Sure, I will try with rabbitmq 3.12. But I'm not sure if any enhancement about locking mechanism in rabbitmq 3.12 to mitigate the cluster split issue. And any configuration suggest ? 
By the way, I deployed rabbitmq cluster by bitnami rabbitmq operator, using their rabbitmq image as well.

**Reply by mkuratczyk** [reply]
Indeed, I don't recall any cluster formation changes, but I believe quite a bit changed in the node boot process, so it could have been fixed accidentally and if not - at least it's easier to reason about the most recent code, without having to recall how things used to work.

Please try the latest 3.12 and collect debug logs when the issue occurs. If you could also try with the non-bitnami image, that would be great to know (I don't suspect this to be the cause but it'd be helpful to definitely rule out this possibility)

**Reply by xixiangzouyibian** [reply]
Thanks Michal, will update the result for 3.12 rabbitmq attempt.
I'm still confused why the queues are broken up into different clusters. Although the cluster split into 2 clusters, one of working clusters should take the whole operations, and all of queues should be created in the same cluster.

**Reply by mkuratczyk** [reply]
If they never formed a 3-node cluster, then you effectively have 2 clusters unaware of each other. If it's possible on the network-level to establish the connection with either one, you'll end up with each having some of the queues/messages.

**Reply by xixiangzouyibian** [reply]
ok, that makes sense.

Shall I configure .spec.podManagementPolicy as Parallel when I try with rabbitmq 3.12 ?

**Reply by mkuratczyk** [reply]
You don't right now? That's the default and in my opinion - the only valid setting for RabbitMQ right now (this may change in 4.0 when we remove Mnesia). But if you see cluster formation problems with `OrderedReady`, that's even more strange.

**Reply by xixiangzouyibian** [reply]
This is why I ask here..
https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#pod-management-policies
> Pod Management Policies
> StatefulSet allows you to relax its ordering guarantees while preserving its uniqueness and identity guarantees via its .spec.podManagementPolicy field.
> 
> OrderedReady Pod Management
> **OrderedReady pod management is the default for StatefulSets**. It implements the behavior described

**Reply by mkuratczyk** [reply]
Cluster Operator sets the value to `Parallel` and for any real deployment, that's what should be used (otherwise if you have a major outage, your cluster may not start again). For testing purposes, if you reproduce the problem with `Parallel`, it's worth checking with `OrderedReady` as well

**Comment by xixiangzouyibian** [comment]
@mkuratczyk I updated rabbitmq to 3.11.23 since this is the version supported by latest rabbitmq operator: https://artifacthub.io/packages/helm/bitnami/rabbitmq-cluster-operator
>     <rabbitmq.operator.chart.version>3.8.7</rabbitmq.operator.chart.version>
>     <rabbitmq.operator.image.version>2.5.0-debian-11-r37</rabbitmq.operator.image.version>
>     <rabbitmq.msgTopology.operator.image.version>1.12.0-debian-11-r58</rabbitmq.msgTopology.operator.image.version>
>     <rabbitmq.image.version>3.11.23-debian-11-r22</rabbitmq.image.version>

The issue still persists (2-node formed a cluster and 1-node formed the other cluster).

**Reply by mkuratczyk** [reply]
This is not the maximum supported version. You can deploy 3.12 with the Operator.

**Comment by xixiangzouyibian** [comment]
@mkuratczyk @Zerpet Hi Team, After nearly 4 days observation, the statistical sample about the issue of cluster formation split is summarized as below ("1" means problematic deployment, "0" means non-problematic deployment):
![image](https://github.com/rabbitmq/cluster-operator/assets/8375855/1132e8c0-20b4-4afe-8301-a21f5fb8e358)

**Almost 10% of the problem probability is very high for a product. Please feel free to find the attachments for further analysis.**

**Reply by mkuratczyk** [reply]
It's 10% in your environment, but 0% in thousands of other environments...

It's not completely clear to me how these log files correlate. Assuming `*2023.10.19-04_59*` are files for 3 nodes from the same attempt, it's worth noting that one of the pods started much later than the other two. It's not clear to me why this would cause a problem (it should join the cluster with the other two nodes), but it could be somehow related to the problem:

```
> head -1 *2023.10.19-04_59*
==> rabbitmq-server-0-2023.10.19-04_59.log <==
rabbitmq 08:57:06.26 

==> rabbitmq-server-1-2023.10.19-04_59.log <==
rabbitmq 09:05:29.51 

==> rabbitmq-server-2-2023.10.19-04_59.log <==
rabbitmq 08:56:46.40 
```

Can you please try:
* the latest 3.12 Bitnami image
* the `rabbitmq:3.12.7-management` image (even if you intend to use the Bitnami image later on, it'd be useful to know if this happens when using the non-bitnami image as well)

And please make sure it's clear how the log files match the attempts. Thanks

**Reply by xixiangzouyibian** [reply]
@mkuratczyk Yes, you can match 3 nodes: rabbitmq-server-0, rabbitmq-server-1, rabbitmq-server-2 with the same timestamp for one attempt, just as you post.

**Reply by xixiangzouyibian** [reply]
ok, trying, but not sure if the rabbitmq operator can be adapted to this image version.

**Reply by mkuratczyk** [reply, marked as answer]
As I said above, you can simply set `image: rabbitmq:3.12.7-management` to deploy 3.12 (with the `rabbitmq` image in this case). There's nothing to "adapt"

**Comment by xixiangzouyibian** [comment]
@mkuratczyk We updated rabbitmq version from bitnami 3.11.23-debian-11-r22 to official 3.12.7-management and no issue found anymore after 3 days monitoring, totally 67 times deployment. We regard the issue has been really resolved. Thanks for your great support.
